File: frontend/packages/workflow/playground/src/components/comment/parsers/json.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-namespace -- namespace is necessary */
import type { CommentEditorBlock } from '../type';
import {
  type CommentEditorBlockFormat,
  CommentEditorDefaultBlocks,
  CommentEditorLeafType,
} from '../constant';

export namespace CommentEditorJSONParser {
  // Processing a single node
  const processNode = (node: CommentEditorBlock): CommentEditorBlock => {
    if ('text' in node && !node.type) {
      return {
        ...node,
        type: CommentEditorLeafType as unknown as CommentEditorBlockFormat,
      };
    }

    if ('type' in node && 'children' in node) {
      return {
        ...node,
        children: (node.children as CommentEditorBlock[]).map(processNode),
      };
    }

    return node as CommentEditorBlock;
  };

  // Main function: handle the entire schema
  const addLeafType = (schema: CommentEditorBlock[]): CommentEditorBlock[] =>
    schema.map(processNode);

  /** Convert JSON to Schema */
  export const from = (value?: string): CommentEditorBlock[] | undefined => {
    if (!value || value === '') {
      return CommentEditorDefaultBlocks as CommentEditorBlock[];
    }
    try {
      const blocks = JSON.parse(value);
      return blocks;
      // eslint-disable-next-line @coze-arch/use-error-in-catch -- no need to handle error
    } catch (error) {
      return;
    }
  };

  /** Schema to JSON */
  export const to = (schema: CommentEditorBlock[]): string | undefined => {
    try {
      const value = JSON.stringify(addLeafType(schema));
      return value;
      // eslint-disable-next-line @coze-arch/use-error-in-catch -- no need to handle error
    } catch (error) {
      return;
    }
  };
}

File: backend/infra/dynconf/impl/static/json.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package static

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"strings"

	"github.com/coze-dev/coze-studio/backend/infra/dynconf/impl/static/internal"
)

type ConfJson struct {
	*internal.RawJson
}

func NewConfJson(rootDir string, groups []string) (*ConfJson, error) {
	jsonFilePath := getJsonPath(rootDir, groups)
	if jsonFilePath == "" {
		return nil, ErrConfigNotExist
	}

	json, err := internal.NewRawJson(jsonFilePath)
	if err != nil {
		return nil, err
	}

	r := &ConfJson{json}
	return r, err
}

func (c *ConfJson) MarshalFunc() MarshalFunc {
	return json.Marshal
}

func (c *ConfJson) UnmarshalFunc() UnmarshalFunc {
	return json.Unmarshal
}

func getJsonPath(rootDir string, groups []string) string {
	var findPaths []string

	for r := len(groups); r > 0; r-- {
		findPaths = append(findPaths,
			filepath.Join(rootDir, fmt.Sprintf("config.%s", strings.Join(groups[:r], "."))))
	}

	findPaths = append(findPaths, filepath.Join(rootDir, "config"))

	for _, path := range findPaths {
		if p, exist := existJsonFile(path); exist {
			return p
		}
	}

	return ""
}

func existJsonFile(fileName string) (string, bool) {
	p := fileName + ".json"
	existed, isDir := internal.FileExist(p)
	if existed && !isDir {
		return p, true
	}

	return "", false
}
